### PR TITLE
core(runner): make LH.Audit.Context immutable

### LINKED ISSUE #9897: Make settings/context immutable
Add a centralized mechanism (e.g deep Object.freeze or readonly type system checks or whatever) to ensure settings immutability.

This issue came up when discussing https://github.com/GoogleChrome/lighthouse/pull/9895.

### PR DESCRIPTION
**Pull request by brendankenny**
**breaking change (see below for details)**
fixes #9897

tl;dr: making this change now allows us to
- trust audits a little less (handy for any model of plugins)
- make audit contexts `postMessage`able, handy for an off-main-thread/extension plugin model if we choose to go in that direction
- pursue the above without having to wait for the 7.0 release, whenever that will be

The change is very small since we only have two lines of code currently modifying an audit context outside of `runner.js` :)

The only real breaking change is that rather than audits pushing to the context to get a top-level LHR warning, they return the warning in their audit `Product`. It's very possible that existing extensions use this functionality. In core, there's only one audit doing so.

Note that `readonly` is still type-checking/advisory level enforcement, so if an audit wants to mess with things it currently can, but that's always been the case and it's useful to leave things open without a reason to do otherwise.

breaking changes:
- audits that previously pushed a string to `context.LighthouseRunWarnings` to get a top-level warning in the LHR will instead need to return a `runWarnings` property on the `Product` return value, set to an array with that string in it. ([example change from core](https://github.com/GoogleChrome/lighthouse/pull/10555/files#diff-2de4b22b2f3e9b5bd6c988afdf375747R184-R215))
- previously audits could push warnings to `context.LighthouseRunWarnings` even if they ended up throwing an error, but that is no longer possible since the run warnings are now part of the normal return value.

### CODE REVIEW

File: lighthouse-core/computed/resource-summary.js [hunk 1]
```diff
@@ -51,7 +51,7 @@ class ResourceSummary {
       'third-party': {count: 0, size: 0},
     };
     const budget = Budget.getMatchingBudget(context.settings.budgets, mainResourceURL);
-    let firstPartyHosts = /** @type {Array<string>} */ ([]);
+    let firstPartyHosts = /** @type {Immutable<Array<string>>} */ ([]);
```
Comment (brendankenny) [inline lighthouse-core/computed/resource-summary.js:54]: this will be able to be just `ReadonlyArray<string>` after we upgrade to tsc 3.7+

File: lighthouse-core/test/runner-test.js [hunk 2]
```diff
@@ -670,31 +670,6 @@ describe('Runner', () => {
     });
   });
 
-  it('includes any LighthouseRunWarnings from errored audits in LHR', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/runner-test.js:673]: this is the sole loss of functionality (being able to push top-level warnings before the audit completely errors out), but I don't think this has ever been behavior anyone cares about or needed

File: types/artifacts.d.ts [hunk 3]
```diff
@@ -509,7 +509,7 @@ declare global {
       export interface MetricComputationDataInput {
         devtoolsLog: DevtoolsLog;
         trace: Trace;
-        settings: Config.Settings;
+        settings: Immutable<Config.Settings>;
```
Comment (brendankenny) [inline types/artifacts.d.ts:512]: this is complicated because this `settings` is just as often not `LH.Config.Settings` as it is, but that's something to sort out on a different day :)

File: types/externs.d.ts [hunk 4]
```diff
@@ -76,6 +76,25 @@ declare global {
...
+    T extends ReadonlyArray<infer R> ? ImmutableArray<R> :
+    T extends Map<infer K, infer V> ? ImmutableMap<K, V> :
+    T extends Set<infer M> ? ImmutableSet<M> :
+    T extends object ? ImmutableObject<T> :
+    T
+
+  // Intermediate immutable interfaces. Prefer e.g. Immutable<Set<T>> over direct use.
+  // TODO: switch to recursive type references when using tsc ≥ 3.7
+  // see https://github.com/microsoft/TypeScript/issues/13923#issuecomment-557509399
+  interface ImmutableArray<T> extends ReadonlyArray<Immutable<T>> {}
```
Comment (brendankenny) [inline types/externs.d.ts:91]: when we upgrade tsc, this can become 
`type ImmutableArray<T> = ReadonlyArray<Immutable<T>>;`, which means things like the built-in `ReadonlyArray<string>` will be transparently switchable with `Immutable<Array<string>>`, giving a lot better ergonomics and functionality

File: lighthouse-core/computed/computed-artifact.js [hunk 5]
```diff
@@ -25,6 +25,8 @@ function makeComputedArtifact(computableArtifact) {
    * @return {ReturnType<C['compute_']>}
    */
   const request = (artifacts, context) => {
+    /** @type {Map<string, ArbitraryEqualityMap>} */
+    // @ts-ignore - break immutability solely for this caching-controller function.
```
Comment (patrickhulce) [inline lighthouse-core/computed/computed-artifact.js:29]: is there a `Mutable<T>` type we could put `Context['computedCache']` through? It'd be nice to not lose that type safety just for the mutability part

File: lighthouse-core/test/runner-test.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/test/runner-test.js:673]: maybe call out the loss of this as a breaking change in PR description so we remember to include that too in final notes?

**Comment by brendankenny** [comment]
> I'm not sure I trust audits any more than I did before, but OK :)

Maybe it would have been more correct to say it gives them better guidance on how to act :)

File: lighthouse-core/test/runner-test.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/test/runner-test.js:673]: > maybe call out the loss of this as a breaking change in PR description so we remember to include that too in final notes?

done

File: lighthouse-core/computed/computed-artifact.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-core/computed/computed-artifact.js:29]: > is there a `Mutable<T>` type we could put `Context['computedCache']` through? It'd be nice to not lose that type safety just for the mutability part

`Map` is a little special (as are `Array` and `Set`) because typescript has a special readonly type for it that removes all the mutating functions from instances, so `Map`s are `ReadonlyMap`s, but `ReadonlyMap`s aren't `Map`s.

`Mutable<T>` could do an unsafe cast for that, but we could just as easily do that here too. There's little enough going on in this line of code that ignoring seemed fine, but maybe the cast would be better?

Comment (patrickhulce) [reply lighthouse-core/computed/computed-artifact.js:29]: not a big deal but even the unsafe cast of ReadonlyMap<T> to Map<T> would be a huge upgrade IMO so we can catch the compatibility of T :)

Comment (brendankenny) [reply lighthouse-core/computed/computed-artifact.js:29]: > so we can catch the compatibility of T

very good point, sold :)

I'll wait for #10461 to be landable so I can use tsc 3.7 stuff

**Comment by brendankenny** [comment]
@connorjclark what should I do about this travis failure?

`legacy-javascript/run.js` ran because this PR touched `legacy-javascript/examine-latest-run.js`, but it was only a mock change, so presumably the diff happened in a different commit/due to dependency changes.
```
0.04s$ git diff --exit-code
diff --git a/lighthouse-core/scripts/legacy-javascript/summary-sizes.txt b/lighthouse-core/scripts/legacy-javascript/summary-sizes.txt
index 2f33adb..9f1e83b 100644
--- a/lighthouse-core/scripts/legacy-javascript/summary-sizes.txt
+++ b/lighthouse-core/scripts/legacy-javascript/summary-sizes.txt
@@ -113,8 +113,8 @@ variants/core-js-2-only-polyfill
    3410 variants/core-js-2-only-polyfill/es6-function-name/main.bundle.min.js
 
 variants/core-js-2-preset-env-esmodules
-3567 total
-2259 variants/core-js-2-preset-env-esmodules/false/main.bundle.min.js
+4121 total
+2813 variants/core-js-2-preset-env-esmodules/false/main.bundle.min.js
 1308 variants/core-js-2-preset-env-esmodules/true/main.bundle.min.js
 
 variants/core-js-3-only-polyfill
@@ -232,13 +232,13 @@ variants/core-js-3-only-polyfill
    3410 variants/core-js-3-only-polyfill/es6-function-name/main.bundle.min.js
 
 variants/core-js-3-preset-env-esmodules
-3567 total
-2259 variants/core-js-3-preset-env-esmodules/false/main.bundle.min.js
+4121 total
+2813 variants/core-js-3-preset-env-esmodules/false/main.bundle.min.js
 1308 variants/core-js-3-preset-env-esmodules/true/main.bundle.min.js
 
 variants/only-plugin
-2793 total
-1152 variants/only-plugin/-babel-plugin-transform-spread/main.bundle.min.js
+3347 total
+1706 variants/only-plugin/-babel-plugin-transform-spread/main.bundle.min.js
  827 variants/only-plugin/-babel-plugin-transform-regenerator/main.bundle.min.js
  814 variants/only-plugin/-babel-plugin-transform-classes/main.bundle.min.js
 
The command "git diff --exit-code" exited with 1.
```

**Comment by brendankenny** [comment]
my local changes from `yarn test-legacy-javascript` are also different than what the travis diff shows, so I'm just going to commit the travis diff and hope the cache hasn't changed in the meantime :)

I'll gate on your approval, though @connorjclark
